File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

// Prototypes
int otherDigit(int oDigit);
void checkType(int digitsNum, long cardNum);

int main(void)
{
    // Prompt for card number
    long cardNum = get_long("Number: ");
    // Initialize a new card number for calculations
    long newCardNum = cardNum;
    
    // Store each digit
    int digit;
    // Initialize the number of digits, the sum in Luhn’s algorithm
    int digitsNum = 0;
    int sum = 0;
    
    for (int i = 1; newCardNum > 0; i++)
    {
        // Get each digit in the card number
        if (newCardNum % 10 == 0)
        {
            digit = 0;
        }
        else
        {
            digit = newCardNum % 10;
        }
        
        // Sum the digits based on their location
        if (i % 2 == 0)
        {
            sum += otherDigit(digit);
        }
        else
        {
            sum += digit;
        }
        
        // New card number after removing the extracted digit
        newCardNum = (newCardNum - digit) / 10;
        
        digitsNum++;
    }
    
    // Check if the first digit in the sum is zero
    if (sum % 10 == 0)
    {
        // If number follows Luhn’s algorithm, check card type
        checkType(digitsNum, cardNum);
    }
    else
    {
        printf("INVALID\n");
    }
}

// Do Luhn’s process on every other digit
int otherDigit(int oDigit)
{
    int multipled = oDigit * 2;
    int firstDigit = multipled % 10;
    int secondDigit = (multipled - firstDigit) / 10;
    
    return firstDigit + secondDigit;
}

// Check card type based on its number of digits and starting digit(s)
void checkType(int digitsNum, long cardNum)
{
    int firstDigit;
    int firstTwoDigits;
    
    if (digitsNum == 15)
    {
        firstTwoDigits = cardNum / pow(10, 13);
        if (firstTwoDigits == 34 || firstTwoDigits == 37)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
        
    }
    else if (digitsNum == 16)
    {
        firstDigit = cardNum / pow(10, 15);
        firstTwoDigits = cardNum / pow(10, 14);
        
        if (firstDigit == 4)
        {
            printf("VISA\n");
        }
        else if (firstTwoDigits == 51 || firstTwoDigits == 52 || firstTwoDigits == 53 || firstTwoDigits == 54 || firstTwoDigits == 55)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (digitsNum == 13)
    {
        firstDigit = cardNum / pow(10, 12);
        if (firstDigit == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
from cs50 import get_int


def main():
    # Prompt for card number, then convert to a string
    card_number = str(get_int('Number: '))

    digits = []
    # Make each digit in the card number an element in a list of integers
    for digit in card_number:
        digits.append(int(digit))
    
    # Reverse the order of the digits to handle its numbers from the right
    digits.reverse()

    sum = 0
    # Sum the digits based on their location
    for i, digit in enumerate(digits):
        # Apply Luhn’s process on every other digit
        if i % 2 == 1:
            digit = other_digits(digit)
        
        sum += digit

    # Initialize the card's type to 'INVALID'
    card_type = 'INVALID'

    # If the card fulfills Luhn’s Algorithm, check its type
    if sum % 10 == 0:
        card_type = check_type(card_number)

    print(card_type)  


def other_digits(digit):
    digit *= 2

    # Extract the first and second digits of the multiplied original
    first_digit = digit % 10
    second_digit = (digit - first_digit) / 10
    
    # Return the sum of the individual two digits
    return first_digit + second_digit


def check_type(card_number):
    # Get the length of and first two digits of the card's number
    length = len(card_number)
    first_digit = int(card_number[0])
    first_two_digits = int(card_number[:2])

    # Return a string of the card's type based on each type's conditions
    if (length == 15) and (first_two_digits in [34, 37]):
        return 'AMEX'
    elif (length in [13, 16]) and (first_digit == 4):
        return 'VISA'
    elif (length == 16) and (first_two_digits in [51, 52, 53, 54, 55]):
        return 'MASTERCARD'
    
    # If the card is not in the avaliable cards, return 'INVALID'
    return 'INVALID'


if __name__ == '__main__':
    main()
